### PR TITLE
Add py-tensorflow as an external package (v2.4.0, CUDA 11.0)

### PR DESCRIPTION
**Pull request by pramodk**
cc: @matz-e : this is just draft for testing!

### CODE REVIEW

**Comment by pramodk** [comment]
CI build failure for pppy-bluepy:

```
[2021-06-17T22:13:24.286Z] ==> No binary for py-bbp-analysis-framework-1.6.42-uvg6dbx3kldoyzpy6afr5ejwehkucvxm found: installing from source
[2021-06-17T22:13:24.557Z] 
[2021-06-17T22:13:24.557Z] gzip: stdin: decompression OK, trailing garbage ignored
[2021-06-17T22:13:24.557Z] /usr/bin/tar: Child returned status 2
[2021-06-17T22:13:24.557Z] /usr/bin/tar: Error is not recoverable: exiting now
[2021-06-17T22:13:24.557Z] 
[2021-06-17T22:13:24.557Z] gzip: stdin: decompression OK, trailing garbage ignored
[2021-06-17T22:13:24.557Z] /usr/bin/tar: Child returned status 2
[2021-06-17T22:13:24.557Z] /usr/bin/tar: Error is not recoverable: exiting now
[2021-06-17T22:13:24.557Z] ==> Using cached archive: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1204/deploy/applications/2021-06-17/.cache/_source-cache/git//nse/bluepy/bluepy-v2.3.0.tar.gz
[2021-06-17T22:13:24.557Z] ==> Warning: Fetching from mirror without a checksum!
[2021-06-17T22:13:24.557Z]   This package is normally checked out from a version control system, but it has been archived on a spack mirror.  This means we cannot know a checksum for the tarball in advance. Be sure that your connection to this mirror is secure!
[2021-06-17T22:13:24.827Z] ==> Using cached archive: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1204/deploy/applications/2021-06-17/.cache/_source-cache/git//nse/bluepy/bluepy-v2.3.0.tar.gz
[2021-06-17T22:13:24.827Z] ==> Warning: Fetching from mirror without a checksum!
[2021-06-17T22:13:24.827Z]   This package is normally checked out from a version control system, but it has been archived on a spack mirror.  This means we cannot know a checksum for the tarball in advance. Be sure that your connection to this mirror is secure!
[2021-06-17T22:13:24.827Z] ==> Error: ProcessError: Command exited with status 2:
[2021-06-17T22:13:24.827Z]     '/usr/bin/tar' '-oxf' '/gpfs/bbp.cscs.ch/ssd/slurmTmpFS/bbprelman/2240542/spack-stage-py-bluepy-2.3.0-ldl3a5is7raxfy2a5v2amgagkuqklkde/bluepy-v2.3.0.tar.gz'
[2021-06-17T22:13:24.827Z] ==> Error: ProcessError: Command exited with status 2:
[2021-06-17T22:13:24.827Z]     '/usr/bin/tar' '-oxf' '/gpfs/bbp.cscs.ch/ssd/slurmTmpFS/bbprelman/2240542/spack-stage-py-bluepy-2.3.0-6ylz5xlt74jnimsddsj23rwn2j4nsi3l/bluepy-v2.3.0.tar.gz'
[2021-06-17T22:13:26.248Z] 
[2021-06-17T22:13:26.248Z] gzip: stdin: decompression OK, trailing garbage ignored
[2021-06-17T22:13:26.248Z] /usr/bin/tar: Child returned status 2
[2021-06-17T22:13:26.248Z] /usr/bin/tar: Error is not recoverable: exiting now
[2021-06-17T22:13:26.248Z] ==> Using cached archive: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jen
```

**Comment by pramodk** [comment]
To test py-tensorflow module (WIP):

```console

# load modules
module use /gpfs/bbp.cscs.ch/apps/hpc/jenkins/pulls/1204/deploy/externals/latest/modules/tcl/linux-rhel7-x86_64
module load unstable
module load py-tensorflow python-dev cuda cudnn 

# allocate gpu node
salloc --partition=prod --account=projX --nodes=1 --time=8:00:00 --exclusive --constraint=volta

kumbhar@r2i3n2:~$ python -c "import tensorflow as tf; tf.test.is_gpu_available()"
2021-06-18 07:26:09.066681: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
WARNING:tensorflow:From <string>:1: is_gpu_available (from tensorflow.python.framework.test_util) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.config.list_physical_devices('GPU')` instead.
2021-06-18 07:26:13.689751: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX512F
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-06-18 07:26:13.710938: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-06-18 07:26:13.714285: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1
2021-06-18 07:26:14.314908: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties:
pciBusID: 0000:1a:00.0 name: Tesla V100-SXM2-16GB computeCapability: 7.0
coreClock: 1.53GHz coreCount: 80 deviceMemorySize: 15.78GiB deviceMemoryBandwidth: 836.37GiB/s
2021-06-18 07:26:14.315931: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 1 with properties:
pciBusID: 0000:1c:00.0 name: Tesla V100-SXM2-16GB computeCapability: 7.0
coreClock: 1.53GHz coreCount: 80 deviceMemorySize: 15.78GiB deviceMemoryBandwidth: 836.37GiB/s
2021-06-18 07:26:14.316853: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 2 with properties:
pciBusID: 0000:88:00.0 name: Tesla V100-SXM2-16GB computeCapability: 7.0
coreClock: 1.53GHz coreCount: 80 deviceMemorySize: 15.78GiB deviceMemoryBandwidth: 836.37GiB/s
2021-06-18 07:26:14.317750: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 3 with properties:
pciBusID: 0000:8a:00.0 name: Tesla V100-SXM2-16GB computeCapability: 7.0
coreClock: 1.53GHz coreCount: 80 deviceMemorySize: 15.78GiB deviceMemoryBandwidth: 836.37GiB/s
2021-06-18 07:26:14.317785: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2021-06-18 07:26:14.323154: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.11
2021-06-18 07:26:14.323227: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublasLt.so.11
2021-06-18 07:26:14.325711: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcufft.so.10
2021-06-18 07:26:14.326935: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcurand.so.10
2021-06-18 07:26:14.331655: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusolver.so.10
2021-06-18 07:26:14.333194: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusparse.so.11
2021-06-18 07:26:14.336658: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.8
2021-06-18 07:26:14.343807: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1862] Adding visible gpu devices: 0, 1, 2, 3
2021-06-18 07:26:14.343852: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2021-06-18 07:26:16.454829: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1261] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-06-18 07:26:16.454898: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1267]      0 1 2 3
2021-06-18 07:26:16.454913: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 0:   N Y Y Y
2021-06-18 07:26:16.454924: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 1:   Y N Y Y
2021-06-18 07:26:16.454935: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 2:   Y Y N Y
2021-06-18 07:26:16.454945: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 3:   Y Y Y N
2021-06-18 07:26:16.460792: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/device:GPU:0 with 14759 MB memory) -> physical GPU (device: 0, name: Tesla V100-SXM2-16GB, pci bus id: 0000:1a:00.0, compute capability: 7.0)
2021-06-18 07:26:16.463431: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/device:GPU:1 with 14759 MB memory) -> physical GPU (device: 1, name: Tesla V100-SXM2-16GB, pci bus id: 0000:1c:00.0, compute capability: 7.0)
2021-06-18 07:26:16.465901: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/device:GPU:2 with 14759 MB memory) -> physical GPU (device: 2, name: Tesla V100-SXM2-16GB, pci bus id: 0000:88:00.0, compute capability: 7.0)
2021-06-18 07:26:16.468651: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/device:GPU:3 with 14759 MB memory) -> physical GPU (device: 3, name: Tesla V100-SXM2-16GB, pci bus id: 0000:8a:00.0, compute capability: 7.0)
```

On non-gpu node:

```console
salloc --partition=prod --account=projX --nodes=1 --time=8:00:00 --exclusive --constraint=cp

kumbhar@r3i7n18:~$ python -c "import tensorflow as tf; print(tf.__version__); tf.test.is_gpu_available()"
2021-06-18 07:42:52.960284: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2.4.0
WARNING:tensorflow:From <string>:1: is_gpu_available (from tensorflow.python.framework.test_util) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.config.list_physical_devices('GPU')` instead.
2021-06-18 07:42:56.422436: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX512F
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-06-18 07:42:56.447892: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-06-18 07:42:56.450607: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1
2021-06-18 07:42:56.832366: E tensorflow/stream_executor/cuda/cuda_driver.cc:328] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2021-06-18 07:42:56.832417: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (r3i7n18): /proc/driver/nvidia/version does not exist
```

@pafonta / @FrancescoCasalegno : could you help to verify if tensorflow module is working as expected? (see above module load commands)

File: deploy/configs/packages.yaml [hunk 1]
```diff
@@ -72,6 +72,12 @@ packages:
       prefix: /usr
   petsc:
     variants: ~fortran
+  py-tensorflow:
+    buildable: false
+    version: [2.4.0]
+    externals:
+    - spec: py-tensorflow cuda_arch=70 ^cuda@11.0.2
+      prefix: /gpfs/bbp.cscs.ch/apps/hpc/external/spack-packages/tensorflow/17-06-2021/tensorflow240_venv_py37_cuda11
```
Comment (matz-e) [inline deploy/configs/packages.yaml]: Dates should be in the other order and adhere to ISO-8601.
```suggestion
      prefix: /gpfs/bbp.cscs.ch/apps/hpc/external/spack-packages/tensorflow/2021-06-17/tensorflow240_venv_py37_cuda11
```

**Review by matz-e — COMMENTED** [review]
Would be good to have some reproducible installation instructions for this.

File: deploy/configs/packages.yaml (hunk 1 above)
Comment (pramodk) [reply deploy/configs/packages.yaml]: I don't think I came across ISO=8601! 😄  I was batting for [RFC 5322 "Internet Message Format"](https://datatracker.ietf.org/doc/rfc5322/?include_text=1) ! 

We are getting rid of external strategy and moving to wheel!

**Comment by pafonta** [comment]
Hello @matz-e,

> @pafonta / @FrancescoCasalegno : could you help to verify if tensorflow module is working as expected? (see above module load commands)

✅ Training a simple neural network on a common dataset has worked.

BB5 setup:
```bash
module use /gpfs/bbp.cscs.ch/apps/hpc/jenkins/pulls/1204/deploy/externals/latest/modules/tcl/linux-rhel7-x86_64
module load unstable
module load py-tensorflow python-dev cuda cudnn 
```

Tensorflow training:
```python
import tensorflow as tf
mnist = tf.keras.datasets.mnist

(x_train, y_train),(x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test, y_test)
```

File: var/spack/repos/builtin/packages/py-tensorflow/package.py [hunk 2]
```diff
@@ -3,752 +3,30 @@
...
-
-        bazel(*args)
-
-        build_pip_package = Executable(
-            'bazel-bin/tensorflow/tools/pip_package/build_pip_package')
-        buildpath = join_path(self.stage.source_path, 'spack-build')
-        build_pip_package('--src', buildpath)
+    depends_on('py-numpy@1.16.0:',  type=('build', 'run'), when='@1.15:')
+    depends_on('cudnn')
+    depends_on('cuda')
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/py-tensorflow/package.py]: This list of dependencies seems a little short. I guess the installation via Pip pulls in a bunch of other stuff. Would be nice to have a longer list here, but in a pinch this should work.

**Review by matz-e — COMMENTED** [review]
Thanks! Conceptually, I think this looks a lot nicer.

**Review by matz-e — DISMISSED** [review]
Fair enough. Good to go in, IMO, if ML sees no issues.

File: var/spack/repos/builtin/packages/py-tensorflow/package.py (hunk 2 above)
Comment (pramodk) [reply var/spack/repos/builtin/packages/py-tensorflow/package.py]: Based on our discussion, I added py-h5py.

**Comment by pramodk** [comment]
Retest this please.

**Review by matz-e — APPROVED** [review]

**Comment by pramodk** [comment]
Tested on BB5. Tomorrow one should be able to do following and use TensorFlow :

```console
module load unstable py-tensorflow python-dev
```

**Pull request merged**